script/job_generator/generate_job_yaml.py

```diff
@@ -70,44 +74,142 @@ def job_prefix_coding(job: Dict[str, Tuple[str, str]]) -> str:
 
 
 @typechecked
-def job_image(job: Dict[str, Tuple[str, str]], container_version: float) -> str:
-    """Generate the image url deppending on the host and device compiler and the selected backend
-    of the job.
+def get_images_from_registry(container_version: str) -> List[str]:
+    """Returns list of all container images for a given tag.
+
+    Args:
+        container_version (str): Container tag
+
+    Returns:
+        List[str]: list of container images
+    """
+    # hard coding the url and project ID is fine, because there is only one
+    # container registry
+    gl = gitlab.Gitlab(url="https://codebase.helmholtz.cloud")
+    project = gl.projects.get("4742")
+    repositories = project.repositories.list(get_all=True)
+
+    # uniform data structure
+    if not isinstance(repositories, list):
+        repositories = [repositories]
+
+    images = []
+    # this process is actual pretty slow
+    # maybe it does a HTML request each time if we iterate over repo.tags.list()
+    # and get a new tag
+    for repo in repositories:
+        for tag in repo.tags.list():
+            if tag.attributes["name"] == container_version:
+                images.append(tag.attributes["location"])
+
+    return images
+
+
+@typechecked
+def job_image(
+    job: Dict[str, Tuple[str, str]],
+    container_version: float,
+    gitlab_images: List[str] = {},
+) -> str:
+    """Generate the image url depending on the host and device compiler and the
+    selected backend of the job.
+
+    If gitlab_images is not empty, only images that are exist are returned.
+    Otherwise, the name of the image is constructed by a set of rules. In this
+    case, the image may not exist. If it looks for an existing image and it does
+    not exist, it returns an image that best matches the software's requirements.
 
     Args:
         job (Dict[str, Tuple[str, str]]): Job dict.
         container_version (float): Container version tag.
+        gitlab_images (List[str]): List of existing container images. If it is
+            empty does not check if image exist.
 
     Returns:
         str: Full container url, which can be used with docker pull.
     """
-    container_url = "registry.hzdr.de/crp/alpaka-group-container/"
-    container_url += "alpaka-ci-ubuntu" + job[UBUNTU][VERSION]
+
+    verified_container_url = [
+        "registry.hzdr.de/crp/alpaka-group-container/"
+        + "alpaka-ci-ubuntu"
+        + job[UBUNTU][VERSION]
+    ]
+
+    is_in_gitlab_images = lambda name: bool(
+        [i for i in gitlab_images if i.startswith(name)]
+    )
+
+    def verify_image(
+        test_url: List[str], verified_url: List[str], gitlab_images: List[str]
+    ) -> bool:
+        """Verify if the test_url is included in gitlab_images.
+
+        Args:
+            test_url (List[str]): Image URL to test. The list will concatenated
+                without whitespaces.
```

Comment:
Is the list separated by whitespace?

Reply:
The list is a python list. I use it, to add words to url at a specific position. For example:

```python
container_url=["registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04", "-gcc"]

# concat without whitespace
print("".join(container_url))
# output: "registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04-gcc"

container_url.insert(1, "-cuda")
# ["registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04", "-cuda", "-gcc"]
print("".join(container_url))
# output: "registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04-cuda-gcc"
```

Reply:
Ah, I see. Then I'd rephrase it like this:

```suggestion
            test_url (List[str]): Image URL to test. The list will be concatenated
                with all whitespaces removed.
```